File: src/main/scala/com/michalplachta/galactic/logic/Followers.scala
package com.michalplachta.galactic.logic

import com.michalplachta.galactic.values.Citizen
import com.michalplachta.galactic.values.Citizen.Stormtrooper

object Followers {
  // PROBLEM #4: clones are counted as followers
  def countFollowersNaive(followers: List[Citizen]): Int = {
    followers.length
  }

  // SOLUTION #4: Traversable + pattern matching
  def countFollowers(followers: List[Citizen]): Int = {
    followers.count {
      case Stormtrooper(_, true) ⇒ false
      case _                     ⇒ true
    }
  }
}

File: src/main/java/com/michalplachta/galactic/java/logic/Followers.java
package com.michalplachta.galactic.java.logic;

import com.michalplachta.galactic.java.values.Citizen;
import io.vavr.collection.List;

import static com.michalplachta.galactic.java.internal.CitizenPatterns.Stormtrooper;
import static io.vavr.API.*;

public class Followers {
    // PROBLEM #4: clones are counted as followers
    public static int countFollowersNaive(List<? extends Citizen> followers) {
        return followers.size();
    }

    // SOLUTION #4: Traversable + pattern matching
    public static int countFollowers(List<? extends Citizen> followers) {
        return followers.count(f -> Match(f).of(
                Case(Stormtrooper($(), $(true)), false),
                Case($(), true)
        ));
    }
}
